### DISCUSSION TITLE
GUI interface to create/edit KiBot configurations

### DISCUSSION THREAD

**Discussion by set-soft** [INTI-CMNB/KiBot#617] [Ideas]
I might be dedicating some time to this. I would like to know:

* What people thinks about it? 
* Any important detail that must be implemented in the GUI?
* Can you collaborate in the implementation? (I'm don't know much about wxWidgets)
* Do you want to donate to the project to accelerate this? (Want to be an sponsor?)
* etc.

I think the value here could be in the fact that you won't need to remember the name of the options.

**Comment by N0cynym** [comment]
Hey @set-soft, this is a great idea and I thought about this last year.

Unfortunately I'm not a good programmer and mostly I work in [egui](https://github.com/emilk/egui) [(rust lang)](https://github.com/rust-lang/rust) projects as maintainer or tester atm.

Great would be, to have a native and web app (which is pretty complex but for the crossed platform kibot it could be a good option because people with different OS use kibot continuously).

The question is if the app should just generate a `.kiplot.yaml` or could at least manipulate `.kicad_sch` and `.kicad_pcb` files!?

Example:

Should setting global variables like `pcb_finish` get just be printed into the `.kiplot.yaml` or could also be written to the PCB file!?

I think this GUI has a lot of benefits and could help none KiCad proffesional to setup the kibot and their kicad project pretty well!

If you wish to collect some ideas for a UI, than I could create some diagrams of my workflow and explain my usage of kibot.

From my side, we use Kibot in over 40 projects now but mostly I need to guide my coworkes with writing the yaml.
I can understand a lot of very proffesional CAD designers that they love the result from kibot but won't struggle with writing a over 500 lines config file for it "from scratch"...
So I created templates for me which I can copy and paste and where I just replace placeholder variables to easily implement and fit kibot in a new project.


Please let me know if someone like me could help at this point.

Thank you for KiBot! 

Best regards from Germany 🇩🇪

**Comment by set-soft** [comment]
> Great would be, to have a native and web app (which is pretty complex but for the crossed platform kibot it could be a good option because people with different OS use kibot continuously).

The web app is interesting, didn't thought about it.
But you don't have to worry about cross platform issues. I plan to try using wxWidgets, used by KiCad, so any GUI stuff will run in the same OSs where KiCad runs.

> The question is if the app should just generate a `.kiplot.yaml` or could at least manipulate `.kicad_sch` and `.kicad_pcb` files!?
> 
> Example:
> 
> Should setting global variables like `pcb_finish` get just be printed into the `.kiplot.yaml` or could also be written to the PCB file!?

Hmmm ... I think preflights are intended to do this kind of modifications. You can config KiBot, then run it and get changes consolidated to the schematic and or PCB.

> I think this GUI has a lot of benefits and could help none KiCad proffesional to setup the kibot and their kicad project pretty well!
> 
> If you wish to collect some ideas for a UI, than I could create some diagrams of my workflow and explain my usage of kibot.

I want to focus on what should be validated and how. But workflows could also help.

> From my side, we use Kibot in over 40 projects now but mostly I need to guide my coworkes with writing the yaml. I can understand a lot of very proffesional CAD designers that they love the result from kibot but won't struggle with writing a over 500 lines config file for it "from scratch"... So I created templates for me which I can copy and paste and where I just replace placeholder variables to easily implement and fit kibot in a new project.

Did you try using parameters as used in the internal templates. I think we need more internal templates to cover the most common outputs.

> Please let me know if someone like me could help at this point.

I think you could take a look at the templates parameters, I now realize that the GUI should offer some mechanism to easily choose some templates.

> Thank you for KiBot!
> 
> Best regards from Germany 🇩🇪

😀

**Reply by N0cynym** [reply]
Great to hear that you plan it to be cross platform.

You're right at this point I just forgot that this happens local. I use kibot in GitLab pipelines so preflights effect the file while the job and artifact lifetime but never change in my case the origin so I forgot that kibot still does this 😅 

I think some diagrams of different workflows could help making the UI perfectly fit for each use case.

This is my GitLab workflow which I did specify:

![Pipeline_Workflows-Kibot GitLab EDA Workflow](https://github.com/INTI-CMNB/KiBot/assets/75754060/45b8c49d-deac-4010-9bf1-f9b4bb90f88e)

At the moment I don't use the full feature set from kibot because I did not test or understand everything (so I use linux tools to do things like replace placeholders, zip or unite files...). But since you moved the documentation from the huge readme to the great [docs page](https://kibot.readthedocs.io/en/master/) I do faster understand the features of kibot.

I hope to offer some templates as well. 
Do you have any wishes how we (kibot useres) should provide kibot configuration templates?

Also do you need some easy graphics which could help beginners to understand kibot?
Like this:
![Pipeline_Workflows-KiCAD Docker](https://github.com/INTI-CMNB/KiBot/assets/75754060/2341992b-d9fa-4f58-b6f5-9e2d09076416)

I do create this graphics in [draw.io](https://app.diagrams.net/) so I could provide a project file if needed...

**Reply by set-soft** [reply]
> Do you have any wishes how we (kibot useres) should provide kibot configuration templates?

Not really. I think that users could try to add a few parameters, like the internal templates provide. Then create common cases with unique names. As an example: the *3DRender* template accepts *_KIBOT_3D_VIEW*, *_KIBOT_ROT_X*, *_KIBOT_ROT_Y* and *_KIBOT_ROT_Z* parameters. So you can easily create different views. But we have *3DRender_bottom*, *3DRender_bottom_straight*, *3DRender_top* and *3DRender_top_straight*, which are common uses. These common cases are implemented using *3DRender*. So users can include *3DRender_top* and get a nice top view without specifying a single parameter.
 
> Also do you need some easy graphics which could help beginners to understand kibot?

I think that what we need are small *how-to*s. Things like: an example of how to solve X or Y problem.

Like in this repo: https://github.com/INTI-CMNB/kibot-examples-1 which renders to this https://inti-cmnb.github.io/kibot-examples-1/value_split/

I know this is an elaborated example, but simple examples could also help.

**Comment by set-soft** [comment]
Hi all!
A basic GUI is now available in the *dev* branch.
Note that advanced stuff, like pre-processing and multiple config files, are partially supported. You edit the consolidated result and the save option saves the consolidated stuff you edited.
Advanced user can still use it to edit/create small portions that are then included by higher level configs.

**Reply by N0cynym** [reply]
Hi @set-soft 

I tried to use the gui but I get following result:

```
 kibot --gui
Traceback (most recent call last):
  File "/usr/bin/kibot", line 33, in <module>
    sys.exit(load_entry_point('kibot==1.8.2', 'console_scripts', 'kibot')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/kibot/__main__.py", line 499, in main
    from .GUI.gui import show_splash
ModuleNotFoundError: No module named 'kibot.GUI'
```

**Reply by N0cynym** [reply]
@set-soft I hope my pull request https://github.com/INTI-CMNB/KiBot/pull/676 fixes it.

**Reply by set-soft** [reply]
I'll apply the patch, but I'll suggest running KiBot *in place*, not installing a *dev* version. 
In this way you can have the stable release installed and run KiBot *dev* on demand.

**Reply by N0cynym** [reply]
I just would like to test the gui in a wsl setup localy unless `1.8.2` is not released.
For my production cases I still use the offical dockers 😉

**Reply by set-soft** [reply]
Ok, the GUI is still beta.
Can you try it as a native Windows application? You should be able to use the GUI with native Window widgets to generate the YAML file, that you can then run using WSL or docker.

**Reply by N0cynym** [reply]
Currently I did build [a debian package](https://github.com/N0cynym/KiBot/actions/runs/11104219173/artifacts/1994867200) for testing in wsl.
If you give me some build advices I will test native builds on windows.

**Reply by N0cynym** [reply]
My last build -> `https://github.com/N0cynym/KiBot/actions/runs/11104219173`

**Reply by set-soft** [reply]
Run a [KiCad command prompt](https://yaqwsx.github.io/KiKit/v1.4/installation/windows/).
Then install KiBot using *pip* (`pip install kibot`). This will install the stable release **and its core dependencies**.
Now you can just clone the *dev* branch of the repo (you can clone the whole repo and then run `git switch dev`)
After this you can run (from the KiCad command prompt) the script named in *src/kibot*. This script runs KiBot from the repo, no need to install.
All operations must be done using the Python shipped with KiCad (this is why you need the *KiCad command prompt*).
From the root of the repo you can call `src/kibot --gui`

**Comment by N0cynym** [comment]
# Translations

My first look into the GUI let's me realise that it is meant for multi languages!?
I there any file/documentation for writing a translation for the GUI content?

**Reply by set-soft** [reply]
wxWidgets support translation for all the available platforms, so yes, you can do translations.
But KiBot isn't currently doing any effort to make it work for its own messages.
The wxWidgets docs are here: https://docs.wxwidgets.org/latest/overview_i18n.html
I'm not sure this is a simple task, we can get most of the UI translated, but tooltips ... they are basically the whole documentation for all the available options.

**Reply by N0cynym** [reply]
I was just thinking about translating the labels.
The translation of the documentation is pretty hard to be done tbh 🙈

**Reply by set-soft** [reply]
All the options docs are embedded in the code.
To translate the message we should wrap them using the _() macro, so the tools used to collect the message can see them.

**Reply by N0cynym** [reply]
Could you provide a macro template so that I just need to update? 
I'm not so common with this in python.

**Reply by set-soft** [reply]
Ok, try translating the following file: https://github.com/INTI-CMNB/KiBot/blob/dev/kibot/locale/de/LC_MESSAGES/kibot.po

The [poEdit](https://poedit.net/) tool is suggested, but I never used it.
